### DISCUSSION TITLE
Should Ibis standardize the position of null values in sorted columns?

### DISCUSSION THREAD

**Discussion by ianmcook** [ibis-project/ibis#7246] [General]
Currently Ibis does not standardize the position of nulls in ordered output. If a user calls `t.order_by(_.x)` or `t.order_by(_.x.desc())` and the column `x` has a mix of null and non-null values, then whether the null values appear at the top (first) or the bottom (last) of the results depends entirely on the default behavior of the Ibis backend you are using. Also, Ibis currently does not expose an option to allow users to control whether nulls sort first or last. 

The behavior of the backend engines is not consistent, but almost all of them do expose an option to control whether nulls sort first or last:

| backend     | default sort position of nulls           | user can specify nulls first or last |
| ----------- | ---------------------------------------- | -------------------------------------|
| bigquery    | first for ascending, last for descending | yes                                  |
| clickhouse  | last                                     | yes                                  |
| dask        | last                                     | yes                                  |
| datafusion  | last for ascending, first for descending | yes                                  |
| druid       | ?                                        | ?                                    |
| duckdb      | last ([since 0.8](https://github.com/duckdb/duckdb/pull/7174)) | yes                                  |
| impala      | last for ascending, first for descending | yes                                  | 
| mssql       | first for ascending, last for descending | no                                   |
| mysql       | first for ascending, last for descending | no                                   |
| oracle      | last for ascending, first for descending | yes                                  |
| pandas      | last                                     | yes                                  |
| polars      | first ([but there's a proposal to change it to last](https://github.com/pola-rs/polars/issues/11063)) | yes                                  |
| postgres    | last for ascending, first for descending | yes                                  |
| pyspark     | first for ascending, last for descending | yes                                  |
| snowflake   | last for ascending, first for descending | yes                                  |
| sqlite      | first for ascending, last for descending | yes                                  |
| trino       | last                                     | yes                                  |

For the backends that do not directly expose the option to control nulls first or nulls last in SQL (mssql, mysql), there is a simple workaround to control this: do `ORDER BY x IS NOT NULL, x` to get nulls first and do `ORDER BY x IS NULL, x` to get nulls last.

Note that this does not consider NaNs in float columns, which are another story and should probably be discussed separately.

**Should Ibis standardize null sort order behavior across all the backends?**

**Comment by cpcloud** [comment]
We have this issue we could resurrect https://github.com/ibis-project/ibis/issues/77

👍🏻 on supporting an API for this

**Reply by ianmcook** [reply]
Cool. FWIW I think it would be strange if Ibis exposed a nulls first / nulls last option without also standardizing one behavior as the default across the backends.

**Reply by cpcloud** [reply]
Yeah, that makes sense. Any thoughts on what the default should be? 🙂

**Reply by ianmcook** [reply]
I think it should be one of:
- last for ascending, first for descending (datafusion, impala, oracle, postgres, snowflake)
- last (clickhouse, dask, duckdb, pandas, trino)

I think it's tough to pick between those two.

**Reply by ianmcook** [reply]
FYI DuckDB changed their behavior from first to last in 0.8: https://github.com/duckdb/duckdb/pull/7174

**Reply by lostmygithubaccount** [reply]
I'm in favor of a `random.random() > .5` call to determine first/last personally

**Reply by saulpw** [reply]
I vote "last" as per my comment in #77: when I ask for the top N in a query, I want to get the top N, and not N nulls.

**Reply by ianmcook** [reply]
+1 to what @saulpw said. The DuckDB folks also argued this:
> `NULLS LAST` is more intuitive when using `ORDER BY` in conjunction with `LIMIT`. With `NULLS FIRST`, the following queries return `NULL` values instead of the desired top-n results:

**Reply by ianmcook** [reply]
Because DuckDB (and maybe other engines?) provides a global option to change the default null sort order, Ibis should probably explicitly specify nulls last even when running queries on engines that default to nulls last.

**Comment by ianmcook** [comment]
A piece of implicit context that is worth making explicit here: _all_ of these engines default to ascending sort order. So Ibis doesn't need to do anything new to standardize ascending/descending.

**Comment by cpcloud** [comment]
Thanks for the discussion all!

I think we'll default to `NULLS LAST`, and provide an API for specifying null ordering in `table.order_by`.

I've created #7252 to track this.
